// typed-structures/dist/collections/queue/queue.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("../..");
const TsQDecorator_1 = require("../../tsQ/decorator/TsQDecorator");
class Queue {
    constructor() {
        this._items = new __1.DoublyLinkedList();
    }
    length() {
        return this._items.length();
    }
    enqueue(item) {
        this._items.push(item);
    }
    dequeue() {
        return this._items.shift();
    }
    peek() {
        return this._items.peek();
    }
    back() {
        return this._items.back();
    }
    empty() {
        return this._items.empty();
    }
    toArray() {
        return this._items.toArray();
    }
}
__decorate([
    TsQDecorator_1.TsQ()
], Queue.prototype, "toArray", null);
exports.Queue = Queue;
