// typed-structures/dist/collections/map/utils/map-entry/map-entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const hash_1 = require("../../../../util/hash");
const flatten_1 = require("../../../../util/flatten");
class MapEntry {
    constructor(key, value) {
        this._value = value;
        this._key = key;
    }
    equals(o) {
        return (this.getKey() === null ? o.getKey() === null : this.getKey() === o.getKey()) &&
            (this.getValue() === null ? o.getValue() === null : this.getValue() === o.getValue());
    }
    getKey() {
        return this._key;
    }
    getValue() {
        return this._value;
    }
    hashCode() {
        let key = typeof this._key === 'object' ? typeof this._key.hashCode === 'function' ? this._key.hashCode() : flatten_1.flattenObject(this._key) : this._key;
        let value = typeof this._value === 'object' ? typeof this._value.hashCode === 'function' ? this._value.hashCode() : flatten_1.flattenObject(this._value) : this._value;
        return hash_1.hash(key + value);
    }
    setValue(value) {
        let old = this.getValue();
        this._value = value;
        return old;
    }
}
exports.MapEntry = MapEntry;
